### PR TITLE
Category model is added

### PR DESCRIPTION
**Pull request by Alexsandr-Yablunovskyi**

### CODE REVIEW

File: src/main/java/yablunovskyi/bookstore/mapper/BookMapper.java [hunk 1]
```diff
@@ -2,21 +2,31 @@
...
 public interface BookMapper {
-    BookDto toDto(Book book);
+    @Mapping(source = "categories", target = "categoriesId")
+    BookResponseDto toDto(Book book);
     
-    Book toBook(CreateBookRequestDto requestDto);
+    Book toBook(BookRequestDto requestDto);
     
-    void updateRequestDtoToBook(CreateBookRequestDto requestDto, @MappingTarget Book book);
+    void updateRequestDtoToBook(BookRequestDto requestDto, @MappingTarget Book book);
```
Comment (AlexVulpine) [inline src/main/java/yablunovskyi/bookstore/mapper/BookMapper.java]: Suggest adding BeanMapping annotation here specifying the nullValuePropertyMappingStrategy property. It will allow you to update only specific fields for book.

File: src/main/java/yablunovskyi/bookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    private final BookService bookService;
+    private final CategoryService categoryService;
+    
+    @ResponseStatus(HttpStatus.CREATED)
+    @Operation(summary = "Create new category in the database",
+            description = "Save a valid information about new category in the database")
+    @PreAuthorize("hasRole('ADMIN')")
+    @PostMapping
+    public CategoryResponseDto createCategory(@RequestBody @Valid CategoryRequestDto requestDto) {
+        return categoryService.save(requestDto);
```
Comment (AlexVulpine) [inline src/main/java/yablunovskyi/bookstore/controller/CategoryController.java:43]: Add ResponseStatus to all your endpoints

**Review by AlexVulpine — APPROVED** [review]
Great job! Left a few remarks for you to consider

File: src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java [hunk 3]
```diff
@@ -15,8 +13,6 @@
 import yablunovskyi.bookstore.security.AuthenticationService;
 import yablunovskyi.bookstore.service.UserService;
 
-@RestController
-@Validated
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java]: Why did you remove RestController annotation?

File: src/main/java/yablunovskyi/bookstore/controller/CategoryController.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    
+    @ResponseStatus
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/controller/CategoryController.java]: this annotation redundant

File: src/main/java/yablunovskyi/bookstore/controller/CategoryController.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    @Operation(summary = "Get all books by category id",
+            description = "Get all books by category id from the database")
+    @PreAuthorize("hasAnyRole('ADMIN', 'USER')")
+    @GetMapping("/{id}/books")
+    public List<BookResponseDtoWithoutCategoryIds> findBooksByCategoryId(
+            @PathVariable @Positive Long id,
+            @RequestBody @Valid Pageable pageable) {
+        return bookService.findBooksByCategoryId(id, pageable);
+    }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/controller/CategoryController.java]:
```suggestion
   
```

File: src/main/java/yablunovskyi/bookstore/dto/book/BookResponseDtoWithoutCategoryIds.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package yablunovskyi.bookstore.dto.book;
+
+import java.math.BigDecimal;
+// use as a response for the endpoint
+// GET: /api/categories/{id}/books (Retrieve books by a specific category)
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/dto/book/BookResponseDtoWithoutCategoryIds.java]: Let’s  remove this comments

File: src/main/java/yablunovskyi/bookstore/model/Book.java [hunk 7]
```diff
@@ -31,6 +36,12 @@ public class Book {
     @Column(nullable = false)
     private String author;
     
+    @ManyToMany
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/model/Book.java:40]: Better exclude from toString()

File: src/main/java/yablunovskyi/bookstore/repository/BookRepository.java [hunk 8]
```diff
@@ -1,9 +1,26 @@
...
-public interface BookRepository extends JpaRepository<Book, Long> {
+public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+    @EntityGraph(attributePaths = "categories")
+    Optional<Book> findById(Long id);
+    
+    @EntityGraph(attributePaths = "categories")
+    Page<Book> findAll(Specification<Book> spec, Pageable pageable);
+    
+    @EntityGraph(attributePaths = "categories")
+    Page<Book> findAll(Pageable pageable);
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/repository/BookRepository.java]: do you really need to get book with categories for each method?  (хороша практика поверти мінімально необхідну кількість данних з БД, підозрюю що деякі з цих методів використовуються для того щоб перемапити їх в дто без категорій, то чи потрібно нам діставати книгу з категоріями в таких випадках?)

File: src/main/resources/db/changelog/changes/06-create-categories-table.yaml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/06-create-categories-table.yaml]: add one empty line (check other files)

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 10]
```diff
@@ -14,3 +14,11 @@ databaseChangeLog:
...
       file: db/changelog/changes/05-set-default-roles.yaml
 
+  - include:
+      file: db/changelog/changes/06-create-categories-table.yaml
+
+  - include:
+      file: db/changelog/changes/07-create-books-categories-table.yaml
+
+
+
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
```

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s  fix a few mistakes;)

File: src/main/resources/db/changelog/changes/06-create-categories-table.yaml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: name
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/06-create-categories-table.yaml:23]: check the max length requirement for your request dto. what could go wrong?

File: src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java (hunk 3 above)
Comment (Alexsandr-Yablunovskyi) [reply src/main/java/yablunovskyi/bookstore/controller/AuthenticationController.java]: By recommendation(. Returned it already

File: src/main/java/yablunovskyi/bookstore/repository/BookRepository.java (hunk 8 above)
Comment (Alexsandr-Yablunovskyi) [reply src/main/java/yablunovskyi/bookstore/repository/BookRepository.java]: I need findById to get book by id and avoid lazyInitException. 
I need findAll(Specification<Book> spec, Pageable pageable) to update book by id
I need findAll(Pageable pageable) to get all books and avoid lazyInitException
I'm sure, there is a way to simplyfied it by I don't find it yet.

File: src/main/java/yablunovskyi/bookstore/dto/book/BookRequestDto.java [hunk 12]
```diff
@@ -6,12 +6,14 @@
...
 import java.math.BigDecimal;
+import java.util.List;
 
-public record CreateBookRequestDto(
+public record BookRequestDto(
         @NotBlank(message = "can't be blank or null")
         String title,
         @NotBlank(message = "can't be blank or null")
         String author,
+        List<Long> categoriesId,
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/dto/book/BookRequestDto.java]:
```suggestion
        @NotEmpty List<@Positive Long> categoriesId,
```

File: src/main/java/yablunovskyi/bookstore/dto/category/CategoryRequestDto.java [hunk 13]
```diff
@@ -0,0 +1,14 @@
+package yablunovskyi.bookstore.dto.category;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+
+public record CategoryRequestDto(
+        @NotBlank(message = "can't be blank or null")
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/dto/category/CategoryRequestDto.java]: specify size?

File: src/main/java/yablunovskyi/bookstore/mapper/BookMapper.java [hunk 14]
```diff
@@ -2,21 +2,31 @@
...
+    
+    BookResponseDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
+    
+    default Long getCategoryId(Category category) {
+        return category.getId();
+    }
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/mapper/BookMapper.java:34]: usually such method can be replaced with `@Mapping(source = "category.id")` 

try experimenting with that

File: src/main/java/yablunovskyi/bookstore/mapper/CategoryMapper.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
+)
+public interface CategoryMapper {
+    CategoryResponseDto toDto(Category category);
+    
+    Category toCategory(CategoryRequestDto requestDto);
+    
+    @BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
+    void updateRequestDtoToCategory(
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/mapper/CategoryMapper.java]:
```suggestion
    void updateCategory(
```

File: src/main/java/yablunovskyi/bookstore/mapper/BookMapper.java (hunk 1 above)
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/mapper/BookMapper.java:28]: makes sense to add:
```
@BeanMapping(nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
```

File: src/main/java/yablunovskyi/bookstore/repository/BookRepository.java [hunk 16]
```diff
@@ -1,9 +1,31 @@
...
 import org.springframework.stereotype.Repository;
 import yablunovskyi.bookstore.model.Book;
 
 @Repository
-public interface BookRepository extends JpaRepository<Book, Long> {
+public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+    @EntityGraph(attributePaths = "categories")
+    Optional<Book> findById(Long id);
+    
+    //for updating
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/repository/BookRepository.java]: such comments are redundant

pls remove

File: src/main/java/yablunovskyi/bookstore/repository/BookRepository.java [hunk 17]
```diff
@@ -1,9 +1,31 @@
...
+    @EntityGraph(attributePaths = "categories")
+    Page<Book> findAll(Specification<Book> spec, Pageable pageable);
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/repository/BookRepository.java]: I don't see any place in the code, where you use this method

File: src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java [hunk 18]
```diff
@@ -1,54 +1,78 @@
...
 public class BookServiceImpl implements BookService {
+    private final CategoryRepository categoryRepository;
     private final BookRepository bookRepository;
     private final BookMapper bookMapper;
     
     @Override
-    public BookDto save(CreateBookRequestDto requestDto) {
+    public BookResponseDto save(BookRequestDto requestDto) {
         Book book = bookMapper.toBook(requestDto);
+        book.setCategories(categoriesIdToCategories(requestDto.categoriesId()));
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java:30]: seems that your method won't work unless you annotate it with `@Transactional`

pls check

File: src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java [hunk 19]
```diff
@@ -1,54 +1,78 @@
...
     }
     
     @Override
-    public BookDto updateById(Long id, CreateBookRequestDto requestDto) {
+    public BookResponseDto updateById(Long id, BookRequestDto requestDto) {
         Book book = bookRepository.findById(id).orElseThrow(
                 () -> new EntityNotFoundException("Can't find a book by id: " + id)
         );
         bookMapper.updateRequestDtoToBook(requestDto, book);
+        book.setCategories(categoriesIdToCategories(requestDto.categoriesId()));
```
Comment (NazarSmith228) [inline src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java:54]: seems that your method won't work unless you annotate it with `@Transactional`

pls check

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java (hunk 19 above)
Comment (Alexsandr-Yablunovskyi) [reply src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java:54]: No, it's working ok

File: src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java (hunk 18 above)
Comment (Alexsandr-Yablunovskyi) [reply src/main/java/yablunovskyi/bookstore/service/impl/BookServiceImpl.java:30]: It's working ok

File: src/main/java/yablunovskyi/bookstore/repository/BookRepository.java [hunk 20]
```diff
@@ -1,9 +1,25 @@
...
+public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+    @EntityGraph(attributePaths = "categories")
+    Optional<Book> findById(Long id);
+    
+    @EntityGraph(attributePaths = "categories")
+    Page<Book> findAll(Pageable pageable);
+    
+    @Query("select b from Book b join fetch b.categories c where c.id = :id")
+    List<Book> findAllByCategoriesId(Long id);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/yablunovskyi/bookstore/repository/BookRepository.java:24]:
```suggestion
```

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊
